include/powerloader/mirror.hpp

```diff
@@ -31,83 +31,132 @@ namespace powerloader
         FAILED
     };
 
+    struct MirrorStats
+    {
+        // Maximum number of allowed parallel connections to this mirror. -1 means no
+        // limit. Dynamically adjusted(decreased) if no fatal(temporary) error will
+        // occur.
+        long allowed_parallel_connections = -1;
+
+        // The maximum number of tried parallel connections to this mirror
+        // (including unsuccessful).
+        int max_tried_parallel_connections = 0;
+
+        // How many transfers from this mirror are currently in progress.
+        int running_transfers = 0;
+
+        // How many transfers was finished successfully from the mirror.
+        int successful_transfers = 0;
+
+        // How many transfers failed.
+        int failed_transfers = 0;
+
+        // Maximum ranges supported in a single request.  This will be automatically
+        // adjusted when mirrors respond with 200 to a range request
+        int max_ranges = 256;
+
+        // Returns the total count of finished transfered.
+        int count_finished_transfers() const
+        {
+            return successful_transfers + failed_transfers;
+        }
+    };
+
     // mirrors should be dict -> urls mapping
     struct POWERLOADER_API Mirror
     {
         Mirror(const Context& ctx, const std::string& url);
-        virtual ~Mirror() = default;
+        virtual ~Mirror();
 
         Mirror(const Mirror&) = delete;
         Mirror& operator=(const Mirror&) = delete;
         Mirror(Mirror&&) = delete;
         Mirror& operator=(Mirror&&) = delete;
 
+        // URL of the mirror
+        const std::string& url() const
+        {
+            return m_url;
+        }
+
+        // Protocol of mirror (can be detected from URL)
+        Protocol protocol() const
+        {
+            return m_protocol;
+        }
+
+        // Statistics about this mirror.
+        // TODO: consider returning by copy for concurrent safety... (like "capturing" the stats at
+        // a given moment - but they might change while observing)
+        const MirrorStats& stats() const
```

Comment:
doesn't this have to be inline to prevent this function to be defined more than once?

Reply:
Every non-static member function of a class which is defined in the class definition is already implicitly inline. (the `inline`s in the `FileIO` type member functions for example are redundant - there might be compiler-specific reasons why you would add `inline` even if it's already implicitly `inline` but that's for not-standard reasons).

I would have needed the `inline` keyword if the definition was outside the class definition, yes.